game/hud/src/components/AnchorIndicator/index.tsx

```diff
@@ -0,0 +1,72 @@
+
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import * as React from 'react';
+import {Position} from '../../services/session/layout';
+
+export interface AnchorIndicatorProps {
+  name: string;     // widget name
+  pos: Position;    // current widget position
+  setAnchor: (name: string, pos: Position) => void;
+}
+
+export interface AnchorIndicatorState {
+}
+
+class AnchorIndicator extends React.Component<AnchorIndicatorProps, AnchorIndicatorState> {
+
+  constructor(props: AnchorIndicatorProps) {
+    super(props);
+  }
+
+  private anchorHere = (event: React.MouseEvent) => {
+    const el: HTMLDivElement = event.currentTarget as HTMLDivElement;
+    function a2n(a: string) : number {
+      switch(a) {
+        case 's': return -1;
+        case 'e': return 1;
+      }
+      return 0;
+    }
+    const pos: Position = Object.assign({}, this.props.pos);
+    let x: number = a2n(el.classList[1][7]);
+    let y: number = a2n(el.classList[2][7]);
+    if (x === pos.anchor.x && y === pos.anchor.y) {
+      pos.anchor.locked = !pos.anchor.locked;
+    } else {
+      pos.anchor.locked = true;
+      pos.anchor.x = x;
+      pos.anchor.y = y;
+    }
+    this.props.setAnchor(this.props.name, pos);
+  }
+
+  render() {
+    const anchors: JSX.Element[] = [];
+    const pos: Position = this.props.pos;
+    let key: number = 0;
+    for (var x = -1; x < 2; x++) {
+      for (let y: number = -1; y < 2; y++) {
+        const classNames: string[] = [ 'anchor' ];
+        classNames.push(x < 0 ? 'x-axis-start' : x > 0 ? 'x-axis-end' : 'x-axis-center');
+        classNames.push(y < 0 ? 'y-axis-start' : y > 0 ? 'y-axis-end' : 'y-axis-center');
+        if (pos.anchor.x === x && pos.anchor.y === y) {
+          classNames.push('current');
+          if (pos.anchor.locked) classNames.push('locked');
```

Comment:
similarly under what circumstances would a widget not have a position or a position without an anchor?  

A widget has to have a position to get onto screen / have a red box to move about.  A position will/should always have an anchor.   To ensure a position always has an anchor I will add some asserts to the savePosition() action so we know that positions entering state have integrity.

Each case should be looked at on merit.  Is there a chance this could be null?  object = localStorage.getItem(...) the answer would be a definite yes, and a null check is a must.  If the answer is not sure, then if there is a sensible default action that can be done, do that, if not, add an assert.  If the answer is no, there is no chance this can be null, no check is required, and in fact adding too much defensive code can mask problems.  

We don't need to defensively code for when shit hits the fan, what we do need to defensively code for is to prevent the fan being hit with shit in the first place.

Some 30+ years ago I inherited some code, much of which was like this:

   void strcpy(char *to, char *from) {
     if (!to || !from) return;
     _strcpy(to, from);
   }

I spent many many hours tracking down bugs being masked by the above code.  That if check should ofc be an assert, but it was too late as there was a code base full of calls to this function (and many others like it) potentially relying on that rather bad behaviour.